Mathlib/CategoryTheory/Sites/NonabelianCohomology/H1.lean

```diff
@@ -0,0 +1,213 @@
+of objects `U : I → C`, we define 1-cochains/1-cocycles/H^1 with values
+in `G` over `U`. (This definition neither requires the assumption that `G`
+is a sheaf, nor that `U` covers the terminal object.)
+As we do not assume that `G` is a presheaf of abelian groups, this
+cohomology theory is only defined in low degrees; in the abelian
+case, it would be a particular case of Čech cohomology (TODO).
+
+## TODO
+
+* show that if `1 ⟶ G₁ ⟶ G₂ ⟶ G₃ ⟶ 1` is a short exact sequence of sheaves
+of groups, and `x₃` is a global section of `G₃` which can be locally lifted
+to a section of `G₂`, there is an associated canonical cohomology class of `G₁`
+which is trivial iff `x₃` can be lifted to a global section of `G₂`.
+(This should hold more generally if `G₂` is a sheaf of sets on which `G₁` acts
+freely, and `G₃` is the quotient sheaf.)
+* deduce a similar result for abelian sheaves
+* when the notion of quasi-coherent sheaves on schemes is defined, show that
+if `0 ⟶ Q ⟶ M ⟶ N ⟶ 0` is an exact sequence of abelian sheaves over a scheme `X`
+and `Q` is the underlying sheaf of a quasi-coherent sheaf, then `M(U) ⟶ N(U)`
+is surjective for any affine open `U`.
+* take the colimit of `OneCohomology G U` over all covering families `U` (for
+a Grothendieck topology)
+
+# References
+
+* [J. Frenkel, *Cohomologie non abélienne et espaces fibrés*][frenkel1957]
+
+-/
+
+universe w' w v u
+
+namespace CategoryTheory
+
+variable {C : Type u} [Category.{v} C]
+
+namespace PresheafOfGroups
+
+variable (G : Cᵒᵖ ⥤ GroupCat.{w}) {X : C} {I : Type w'} (U : I → C)
+
+/-- A zero cochain consists of a family of sections. -/
+def ZeroCochain := ∀ (i : I), G.obj (Opposite.op (U i))
+
+instance : Group (ZeroCochain G U) := by
+  dsimp [ZeroCochain]
+  infer_instance
+
+namespace Cochain₀
+
+@[simp]
+lemma one_apply (i : I) : (1 : ZeroCochain G U) i = 1 := rfl
+
+@[simp]
+lemma inv_apply (γ : ZeroCochain G U) (i : I) : γ⁻¹ i = (γ i)⁻¹ := rfl
+
+@[simp]
+lemma mul_apply (γ₁ γ₂ : ZeroCochain G U) (i : I) : (γ₁ * γ₂) i = γ₁ i * γ₂ i := rfl
+
+end Cochain₀
+
+/-- A 1-cochain of a presheaf of groups `G : Cᵒᵖ ⥤ GroupCat` on a family `U : I → C` of objects
+consists of the data of an element in `G.obj (Opposite.op T)` whenever we have elements
+`i` and `j` in `I` and maps `a : T ⟶ U i` and `b : T ⟶ U j`, and it must satisfy a compatibility
+with respect to precomposition. (When the binary product of `U i` and `U j` exists, this
+data for all `T`, `a` and `b` corresponds to the data of a section of `G` on this product.) -/
+@[ext]
+structure OneCochain where
+  /-- the data involved in a 1-cochain -/
+  ev (i j : I) ⦃T : C⦄ (a : T ⟶ U i) (b : T ⟶ U j) : G.obj (Opposite.op T)
+  ev_precomp (i j : I) ⦃T T' : C⦄ (φ : T ⟶ T') (a : T' ⟶ U i) (b : T' ⟶ U j) :
+    G.map φ.op (ev i j a b) = ev i j (φ ≫ a) (φ ≫ b) := by aesop
+
+namespace OneCochain
+
+attribute [simp] OneCochain.ev_precomp
+
+instance : One (OneCochain G U) where
+  one := { ev := fun _ _ _ _ _ => 1 }
```

Comment:
```suggestion
  one := { ev := fun _ _ _ _ _ ↦ 1 }
```
I thought mathlib preferred `mapsto`. Do you not like it? I don't really know the style here, maybe both are fine?